### PR TITLE
Reuse LeafReader where possible in StoredRowLookup

### PR DESCRIPTION
**Pull request by romseygeek**
Currently StoredRowLookup will pull new iterators on every new document,
but this will quite often not be necessary.  Instead, we compare the previously
cached reader and document id against the new lookup request, and if the
reader is the same and the document id is higher, we don't reset readers on
any internal expressions so that iterators are re-used.

### CODE REVIEW

**Comment by romseygeek** [comment]
Benchmarks are happy:
```
compare_run.py --v1 branch:master --v2 branch:aw/storage/stored-row-always-in-order --spec specs/queries.toml
```
```
# Results (server side duration in ms)
V1: 5.10.0-a779810d48892a6ad0f5a5b86ae7f69717c42f52
V2: 5.10.0-c107c6204382a29cee78024ea14c49ffd8a7643b

Q: select * from uservisits limit 10
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        0.515 ±    0.899 |      0.304 |      0.435 |      0.555 |     28.371 |
|   V2    |        0.505 ±    0.895 |      0.290 |      0.421 |      0.550 |     28.204 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -   1.88%                           -   3.24%
There is a 18.93% probability that the observed difference is not random, and the best estimate of that difference is 1.88%
The test has no statistical significance

Q: select * from uservisits limit 100
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        0.605 ±    0.121 |      0.501 |      0.581 |      0.624 |      2.144 |
|   V2    |        0.489 ±    0.098 |      0.372 |      0.475 |      0.516 |      1.994 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -  21.17%                           -  20.20%
There is a 100.00% probability that the observed difference is not random, and the best estimate of that difference is 21.17%
The test has statistical significance

Q: select * from uservisits limit 100
C: 20
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        0.989 ±    0.954 |      0.527 |      0.804 |      1.012 |     12.520 |
|   V2    |        0.571 ±    0.602 |      0.364 |      0.416 |      0.571 |      8.498 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -  53.58%                           -  63.50%
There is a 100.00% probability that the observed difference is not random, and the best estimate of that difference is 53.58%
The test has statistical significance

Q: select * from uservisits limit 1000
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        3.979 ±    0.597 |      3.690 |      3.902 |      3.974 |     13.498 |
|   V2    |        2.629 ±    0.254 |      2.434 |      2.602 |      2.651 |      6.945 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -  40.86%                           -  39.95%
There is a 100.00% probability that the observed difference is not random, and the best estimate of that difference is 40.86%
The test has statistical significance

Q: select * from uservisits limit 10000
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |       38.254 ±    1.123 |     37.454 |     37.960 |     38.194 |     50.537 |
|   V2    |       25.374 ±    1.740 |     24.680 |     25.109 |     25.266 |     49.915 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -  40.49%                           -  40.75%
There is a 100.00% probability that the observed difference is not random, and the best estimate of that difference is 40.49%
The test has statistical significance

Q: select * from uservisits limit 100000
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |      424.161 ±   18.115 |    378.174 |    423.107 |    436.444 |    503.178 |
|   V2    |      269.404 ±   12.998 |    244.273 |    272.976 |    278.115 |    328.567 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -  44.63%                           -  43.14%
There is a 100.00% probability that the observed difference is not random, and the best estimate of that difference is 44.63%
The test has statistical significance

Q: select * from uservisits order by "adRevenue" limit 10
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        6.756 ±    1.617 |      6.339 |      6.563 |      6.654 |     53.688 |
|   V2    |        7.023 ±    1.840 |      6.395 |      6.847 |      6.997 |     61.084 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               +   3.88%                           +   4.24%
There is a 99.94% probability that the observed difference is not random, and the best estimate of that difference is 3.88%
The test has statistical significance

Q: select * from uservisits order by "adRevenue" limit 10
C: 20
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |       57.438 ±   21.885 |     15.535 |     54.861 |     73.634 |    157.223 |
|   V2    |       55.448 ±   20.829 |      9.656 |     51.668 |     70.342 |    125.894 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -   3.53%                           -   5.99%
There is a 96.26% probability that the observed difference is not random, and the best estimate of that difference is 3.53%
The test has no statistical significance

Q: select "cCode", count(*) from uservisits group by 1 order by 2 desc limit 20
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        8.842 ±    8.796 |      7.820 |      7.952 |      8.098 |    177.243 |
|   V2    |        8.912 ±   11.515 |      7.847 |      8.001 |      8.107 |    225.726 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               +   0.79%                           +   0.62%
There is a 8.60% probability that the observed difference is not random, and the best estimate of that difference is 0.79%
The test has no statistical significance

Q: select extract(day from "visitDate"), count(*) from uservisits group by 1 order by 2 desc limit 20
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |       50.235 ±    6.515 |     48.851 |     49.839 |     50.050 |    193.140 |
|   V2    |       47.707 ±    5.782 |     44.665 |     47.364 |     48.922 |    163.839 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -   5.16%                           -   5.09%
There is a 100.00% probability that the observed difference is not random, and the best estimate of that difference is 5.16%
The test has statistical significance


System/JVM Metrics (durations in ms, byte-values in MB)
    |    YOUNG GC            |       OLD GC           |      HEAP         |     ALLOC
    |  cnt      avg      max |  cnt      avg      max |  initial     used |     rate      total
 V1 | 1771     1.76     1.11 |  534    15.57    40.32 |      268       53 |  2526.58     326513
 V2 |  631     3.06     1.68 |  204    19.24    16.11 |      268       61 |   837.43     120240

Top allocation frames
  V1
    Unsafe.allocateUninitializedArray(Class, int) total=109412645112, count=7095
    MappedMemorySegmentImpl.dup(...) total=41035588086, count=2816
    MemorySegmentIndexInput.buildSlice(...) total=27629841896, count=2050
    Arrays.copyOf(...) total=26546001616, count=1189
    Array.newInstance(Class, int) total=12810080623, count=799
    MemorySegmentIndexInput$SingleSegmentImpl.<init>(...) total=11663191456, count=711
    DirectMonotonicReader.getInstance(...) total=11532352313, count=782
    StringConcatHelper.newString(...) total=10967720064, count=785
    CustomLucene90DocValuesProducer$BaseSortedDocValues.termsEnum() total=8713653080, count=447
    StringUTF16.compress(...) total=7818580552, count=475
  V2
    Arrays.copyOf(...) total=27520270048, count=2840
    StringUTF16.compress(...) total=14740106096, count=2122
    BytesRef.utf8ToString() total=9791344432, count=1772
    ArrayList.iterator() total=7376982696, count=278
    HashMap.newNode(...) total=7132887640, count=1325
    AbstractBigArray.newBytePage(int) total=6178410736, count=1037
    CharsRef.toString() total=5222230864, count=763
    DirectMethodHandle.allocateInstance(Object) total=4952664880, count=3144
    FetchCollector.collect(...) total=4138346804, count=9
    IndexSearcher.getNumClausesCheckVisitor() total=3535350808, count=132

Top frames (by count)
  V1
    Unsafe.allocateUninitializedArray(Class, int) total=109412645112, count=7095
    GregorianChronology.isLeapYear(int) total=4860, count=4860
    MappedMemorySegmentImpl.dup(...) total=41035588086, count=2816
    MemorySegmentIndexInput.buildSlice(...) total=27629841896, count=2050
    TopFieldCollector$SimpleFieldCollector$1.collect(int) total=1985, count=1985
    Long.valueOf(long) total=810488436, count=1706
    Row.materialize() total=333482672, count=1295
    DirectMethodHandle.allocateInstance(Object) total=2207332881, count=1258
    IntObjectHashMap.allocateBuffers(int) total=676166968, count=1191
    Arrays.copyOf(...) total=26546001616, count=1189
  V2
    Long.valueOf(long) total=2171152282, count=3965
    GregorianChronology.isLeapYear(int) total=3443, count=3443
    Row.materialize() total=641548608, count=3292
    DirectMethodHandle.allocateInstance(Object) total=4952664880, count=3144
    Arrays.copyOf(...) total=27520270048, count=2840
    IntObjectHashMap.allocateBuffers(int) total=1015578131, count=2146
    StringUTF16.compress(...) total=14740106096, count=2122
    BytesRef.utf8ToString() total=9791344432, count=1772
    TopFieldCollector$SimpleFieldCollector$1.collect(int) total=1701, count=1701
    HashMap.newNode(...) total=7132887640, count=1325
```

File: server/src/main/java/io/crate/expression/reference/doc/lucene/StoredRowLookup.java [hunk 1]
```diff
@@ -238,13 +239,15 @@ private void registerRef(Reference ref, boolean fromParents) {
...
+        protected void moveToDoc(boolean reuseReader) throws IOException {
             fieldsVisitor.reset();
             this.docVisited = false;
             this.parsedSource = null;
             for (var expr : expressions) {
                 // TODO cache reader context and ensure docs are in-order
-                expr.expression.setNextReader(readerContext);
+                if (reuseReader == false) {
+                    expr.expression.setNextReader(readerContext);
+                }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/reference/doc/lucene/StoredRowLookup.java:250]: Doesn't this warrant removing the TODO?

**Review by mfussenegger — APPROVED** [review]
> Benchmarks are happy:

Could you also post results compared to 5.9.x to see how far off we are?

**Comment by romseygeek** [comment]
Looks like we're back to where we were in 5.9, or even slightly ahead:
```
compare_run.py --v1 branch:5.9 --v2 branch:aw/storage/stored-row-always-in-order --spec specs/queries.toml
```
```
# Results (server side duration in ms)
V1: 5.9.3-c5619e50036acb5ef8b43c6b453f08d6480c4737
V2: 5.10.0-c107c6204382a29cee78024ea14c49ffd8a7643b

Q: select * from uservisits limit 10
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        0.587 ±    1.676 |      0.288 |      0.430 |      0.609 |     52.360 |
|   V2    |        0.515 ±    0.885 |      0.304 |      0.435 |      0.566 |     27.899 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -  12.97%                           +   1.14%
There is a 76.71% probability that the observed difference is not random, and the best estimate of that difference is 12.97%
The test has no statistical significance

Q: select * from uservisits limit 100
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        0.522 ±    0.264 |      0.405 |      0.501 |      0.542 |      7.952 |
|   V2    |        0.488 ±    0.552 |      0.375 |      0.448 |      0.492 |     17.621 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -   6.83%                           -  11.10%
There is a 92.50% probability that the observed difference is not random, and the best estimate of that difference is 6.83%
The test has no statistical significance

Q: select * from uservisits limit 100
C: 20
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        0.936 ±    1.009 |      0.387 |      0.710 |      0.948 |     12.995 |
|   V2    |        0.604 ±    0.674 |      0.380 |      0.480 |      0.613 |     12.658 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -  43.16%                           -  38.53%
There is a 100.00% probability that the observed difference is not random, and the best estimate of that difference is 43.16%
The test has statistical significance

Q: select * from uservisits limit 1000
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        2.651 ±    0.169 |      2.489 |      2.628 |      2.679 |      5.345 |
|   V2    |        2.628 ±    0.280 |      2.442 |      2.602 |      2.661 |      9.356 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -   0.86%                           -   0.98%
There is a 97.08% probability that the observed difference is not random, and the best estimate of that difference is 0.86%
The test has no statistical significance

Q: select * from uservisits limit 10000
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |       25.561 ±    0.845 |     25.123 |     25.396 |     25.492 |     34.402 |
|   V2    |       25.687 ±    0.740 |     25.091 |     25.586 |     25.721 |     35.797 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               +   0.49%                           +   0.75%
There is a 98.82% probability that the observed difference is not random, and the best estimate of that difference is 0.49%
The test has statistical significance

Q: select * from uservisits limit 100000
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |      277.212 ±   13.734 |    253.710 |    278.523 |    284.984 |    361.252 |
|   V2    |      274.203 ±   12.923 |    249.741 |    278.007 |    282.875 |    329.361 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -   1.09%                           -   0.19%
There is a 98.80% probability that the observed difference is not random, and the best estimate of that difference is 1.09%
The test has statistical significance

Q: select * from uservisits order by "adRevenue" limit 10
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        7.877 ±    2.474 |      6.288 |      7.857 |      7.971 |     83.242 |
|   V2    |        6.827 ±    1.812 |      6.371 |      6.624 |      6.754 |     59.196 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -  14.29%                           -  17.03%
There is a 100.00% probability that the observed difference is not random, and the best estimate of that difference is 14.29%
The test has statistical significance

Q: select * from uservisits order by "adRevenue" limit 10
C: 20
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |       61.017 ±   23.525 |     15.377 |     56.952 |     78.115 |    134.964 |
|   V2    |       57.197 ±   21.629 |     14.745 |     53.434 |     71.825 |    131.615 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -   6.46%                           -   6.37%
There is a 99.98% probability that the observed difference is not random, and the best estimate of that difference is 6.46%
The test has statistical significance

Q: select "cCode", count(*) from uservisits group by 1 order by 2 desc limit 20
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        8.942 ±   13.555 |      7.600 |      7.857 |      7.993 |    255.504 |
|   V2    |        9.439 ±   13.541 |      7.998 |      8.167 |      8.278 |    266.831 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               +   5.41%                           +   3.87%
There is a 43.84% probability that the observed difference is not random, and the best estimate of that difference is 5.41%
The test has no statistical significance

Q: select extract(day from "visitDate"), count(*) from uservisits group by 1 order by 2 desc limit 20
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |       51.589 ±    5.976 |     47.553 |     51.071 |     52.275 |    178.037 |
|   V2    |       51.682 ±    5.290 |     50.220 |     51.249 |     51.703 |    164.981 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               +   0.18%                           +   0.35%
There is a 20.61% probability that the observed difference is not random, and the best estimate of that difference is 0.18%
The test has no statistical significance


System/JVM Metrics (durations in ms, byte-values in MB)
    |    YOUNG GC            |       OLD GC           |      HEAP         |     ALLOC
    |  cnt      avg      max |  cnt      avg      max |  initial     used |     rate      total
 V1 | 1005     2.35     0.62 |  264    18.32    15.92 |      268      204 |  1373.56     204327
 V2 |  631     3.15     1.37 |  200    20.02    21.60 |      268       61 |   802.63     117752

Top allocation frames
  V1
    Long.valueOf(long) total=48094539057, count=12303
    Arrays.copyOf(...) total=26802630712, count=2911
    StringUTF16.compress(...) total=19921801504, count=2035
    HashMap.newNode(...) total=14917003875, count=2056
    StoredFieldVisitor.binaryField(...) total=8462133789, count=822
    ArrayList.iterator() total=5963997200, count=153
    HashMap.resize() total=5891594830, count=731
    AbstractBigArray.newBytePage(int) total=5880782688, count=670
    CharsRef.toString() total=4874141433, count=543
    ByteSourceJsonBootstrapper.constructParser(...) total=4816119579, count=680
  V2
    Arrays.copyOf(...) total=25708140792, count=3142
    StringUTF16.compress(...) total=15191736936, count=2243
    BytesRef.utf8ToString() total=9964995000, count=1756
    HashMap.newNode(...) total=8485091936, count=1404
    ArrayList.iterator() total=6641830128, count=162
    CharsRef.toString() total=6607416568, count=754
    AbstractBigArray.newBytePage(int) total=5252192320, count=830
    DirectMethodHandle.allocateInstance(Object) total=3670970072, count=564
    FetchCollector.collect(...) total=3519795212, count=10
    Arrays.copyOfRange(...) total=3474241777, count=616

Top frames (by count)
  V1
    Long.valueOf(long) total=48094539057, count=12303
    Row.materialize() total=633068408, count=3701
    GregorianChronology.isLeapYear(int) total=3478, count=3478
    Arrays.copyOf(...) total=26802630712, count=2911
    TopFieldCollector$SimpleFieldCollector$1.collect(int) total=2582, count=2582
    IntObjectHashMap.allocateBuffers(int) total=1060584088, count=2297
    HashMap.newNode(...) total=14917003875, count=2056
    StringUTF16.compress(...) total=19921801504, count=2035
    GroupByOptimizedIterator.aggregateValues(...) total=869, count=869
    GroupingCollector.iter(Map, Row) total=861, count=861
  V2
    Long.valueOf(long) total=2210589905, count=5481
    Row.materialize() total=655268336, count=4325
    GregorianChronology.isLeapYear(int) total=4248, count=4248
    Arrays.copyOf(...) total=25708140792, count=3142
    IntObjectHashMap.allocateBuffers(int) total=994990536, count=2320
    StringUTF16.compress(...) total=15191736936, count=2243
    BytesRef.utf8ToString() total=9964995000, count=1756
    TopFieldCollector$SimpleFieldCollector$1.collect(int) total=1717, count=1717
    HashMap.newNode(...) total=8485091936, count=1404
    MemorySessionImpl.checkValidStateRaw() total=976, count=976
```

**Pull request merged**